From 270f229a798e911d5933d1f884c05e4e7404f99e Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Sat, 25 Jul 2026 01:26:24 -0400
Subject: [PATCH 234/268] amdxc: Implement IAmdExtD3DCreateDevice.

Used by native AGS.
---
 dlls/amdxc64/amdxc_interfaces.idl | 37 +++++++++++++++
 dlls/amdxc64/main.c               | 75 +++++++++++++++++++++++++++++++
 2 files changed, 112 insertions(+)

diff --git a/dlls/amdxc64/amdxc_interfaces.idl b/dlls/amdxc64/amdxc_interfaces.idl
index 9fe6aa5e3bb..5f40be6b58a 100644
--- a/dlls/amdxc64/amdxc_interfaces.idl
+++ b/dlls/amdxc64/amdxc_interfaces.idl
@@ -213,6 +213,32 @@ typedef struct
     const char *kernelName;
 } AmdExtD3DPipelineCrossCompileInfo;
 
+typedef enum
+{
+    AmdExtD3DStructTypeAppRegId         = 0x00000001,
+    AmdExtD3DStructTypeDisableDevDriver = 0x00000002,
+    AmdExtD3DStructTypeWorkTiling       = 0x00000003,
+    AmdExtD3DStructTypePeerUmdOnDxgi    = 0x00000004,
+} AmdExtD3DCreateDeviceStructTypeId;
+
+typedef struct
+{
+    AmdExtD3DCreateDeviceStructTypeId   type;
+    struct AmdExtD3DCreateDeviceInfo*   pNext;
+} AmdExtD3DCreateDeviceInfo;
+
+typedef struct
+{
+    AmdExtD3DCreateDeviceInfo info;
+    const WCHAR *appName;
+    UINT appVersion;
+    const WCHAR *pEngineName;
+    UINT engineVersion;
+} AmdExtAppRegInfo;
+
+cpp_quote("#define AMD_MAKE_VERSION(major,minor,patch) ((major << 22) | (minor << 12) | patch)")
+cpp_quote("#define AMD_UNSPECIFIED_VERSION 0xFFFFAD00")
+
 [
     object,
     uuid(E6144584-03DE-439C-9C0B-43AE6D009BC6),
@@ -327,3 +353,14 @@ interface IAmdExtD3DDevice8 : IAmdExtD3DDevice7
 {
     HRESULT GetWaveMatrixProperties([in, out] SIZE_T *pCount, [in, out] AmdExtWaveMatrixProperties *pProperties);
 }
+
+[
+    object,
+    uuid(C4200955-FD57-481A-BD22-947174C6C698),
+    local
+]
+interface IAmdExtD3DCreateDevice : IUnknown
+{
+    HRESULT AmdD3D12CreateDevice(IDXGIAdapter *adapter, D3D_FEATURE_LEVEL minLevel,
+                                 REFIID iid, void **device, AmdExtD3DCreateDeviceInfo* ext);
+}
diff --git a/dlls/amdxc64/main.c b/dlls/amdxc64/main.c
index 04047583754..3d685bea467 100644
--- a/dlls/amdxc64/main.c
+++ b/dlls/amdxc64/main.c
@@ -638,6 +638,78 @@ static const struct AmdExtD3DFactory amd_d3d_factory = {
     .IAmdExtD3DFactory_iface = { &AmdExtD3DFactory_vtable },
 };
 
+struct AmdExtD3DCreateDevice
+{
+    IAmdExtD3DCreateDevice IAmdExtD3DCreateDevice_iface;
+};
+
+HRESULT STDMETHODCALLTYPE AmdExtD3DCreateDevice_QueryInterface(IAmdExtD3DCreateDevice *iface, REFIID iid, void **out)
+{
+    FIXME("%p %s %p stub!\n", iface, debugstr_guid(iid), out);
+    return E_NOTIMPL;
+}
+
+ULONG STDMETHODCALLTYPE AmdExtD3DCreateDevice_AddRef(IAmdExtD3DCreateDevice *iface)
+{
+    return 2;
+}
+
+ULONG STDMETHODCALLTYPE AmdExtD3DCreateDevice_Release(IAmdExtD3DCreateDevice *iface)
+{
+    return 1;
+}
+
+HRESULT STDMETHODCALLTYPE AmdExtD3DCreateDevice_AmdD3D12CreateDevice(IAmdExtD3DCreateDevice *iface, IDXGIAdapter *adapter, D3D_FEATURE_LEVEL minLevel,
+                                                                     REFIID iid, void **device, AmdExtD3DCreateDeviceInfo *ext)
+{
+    HRESULT ret;
+    HMODULE d3d12 = LoadLibraryW(L"d3d12.dll");
+    static typeof(D3D12CreateDevice) *pD3D12CreateDevice;
+
+    TRACE("%p %p %u %s %p %p\n", iface, adapter, minLevel, debugstr_guid(iid), device, ext);
+
+    if (!d3d12) return E_FAIL;
+
+    pD3D12CreateDevice = (void *)GetProcAddress(d3d12, "D3D12CreateDevice");
+    if (!pD3D12CreateDevice) return E_FAIL;
+
+    ret = pD3D12CreateDevice((IUnknown *)adapter, minLevel, iid, device);
+
+    while (ext)
+    {
+        switch (ext->type)
+        {
+            case AmdExtD3DStructTypeAppRegId:
+            {
+                AmdExtAppRegInfo *info = (void *)ext;
+                /* TODO add an interface into vkd3d-proton for this */
+                TRACE("reg info %s %u %s %u\n", debugstr_w(info->appName), info->appVersion,
+                                                debugstr_w(info->pEngineName), info->engineVersion);
+                break;
+            }
+            default:
+                FIXME("Unimplemented ext type %u\n", ext->type);
+                break;
+        }
+
+        ext = (void *)ext->pNext;
+    }
+
+    FreeLibrary(d3d12);
+    return ret;
+}
+
+static const struct IAmdExtD3DCreateDeviceVtbl AmdExtD3DCreateDevice_vtable = {
+    AmdExtD3DCreateDevice_QueryInterface,
+    AmdExtD3DCreateDevice_AddRef,
+    AmdExtD3DCreateDevice_Release,
+    AmdExtD3DCreateDevice_AmdD3D12CreateDevice,
+};
+
+static const struct AmdExtD3DCreateDevice amd_d3d_create = {
+    .IAmdExtD3DCreateDevice_iface = { &AmdExtD3DCreateDevice_vtable },
+};
+
 HRESULT CDECL AmdExtD3DCreateInterface(IUnknown *outer, REFIID iid, void **obj)
 {
     TRACE("outer %p, iid %s, obj %p\n", outer, debugstr_guid(iid), obj);
@@ -656,6 +728,9 @@ HRESULT CDECL AmdExtD3DCreateInterface(IUnknown *outer, REFIID iid, void **obj)
     } else if (IsEqualGUID(iid, &IID_IAmdExtD3DFactory)) {
         *obj = (void *)&amd_d3d_factory.IAmdExtD3DFactory_iface;
         return S_OK;
+    } else if (IsEqualGUID(iid, &IID_IAmdExtD3DCreateDevice)) {
+        *obj = (void *)&amd_d3d_create.IAmdExtD3DCreateDevice_iface;
+        return S_OK;
     } else {
         FIXME("unknown guid: %s\n", debugstr_guid(iid));
     }
-- 
2.55.0

